File: 2020/01.rb
numbers = File.readlines('01_1_input.txt').map(&:to_i)

# naive
numbers.each do |m|
  n = numbers.reduce(nil) do |_ ,i|
    o = numbers.find { m + i + _1 == 2020 }
    break i * o if o
  end
  break m * n if n
end

# sort + binary search
sorted_numbers = numbers.sort
sorted_numbers.each do |m|
  n = numbers.reduce(nil) do |_ ,i|
    o = sorted_numbers.bsearch { m + i + _1 - 2000 }

    break i * o if o
  end

  break m * n if n
end

File: 2020/01.ml
#use "helpers.ml";;

let numbers = read_lines "01_input.txt" |> List.map int_of_string;;


(* Part 1 *)
List.find_map (fun m ->
    List.find_map (fun n ->
        match m + n with
        | 2020 -> Some(m * n)
        | _ -> None
      ) numbers
  ) numbers;;
(* - : int = 1016619 *)


(* Part 2 *)
List.find_map (fun m ->
    List.find_map (fun n ->
        List.find_map (fun o ->
            match n + m + o with
            | 2020 -> Some(n * m * o)
            | _ -> None
          ) numbers
      ) numbers
  ) numbers;;
(*  - : int = 218767230 *)
